### PR TITLE
Add E2E tests for diagnostic and formatting

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR adds end-to-end tests for diagnostics and formatting capability of the Ruff VS Code extension.

### Resources

* Testing the language server: https://code.visualstudio.com/api/language-extensions/language-server-extension-guide#testing-the-language-server
* Testing extension: https://code.visualstudio.com/api/working-with-extensions/testing-extension
* LSP sample VS Code extension: https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-sample

## Test Plan

Run `just e2e-test` or `npm run test`

Preview:

https://github.com/astral-sh/ruff-vscode/assets/67177269/ae0d214f-e858-4558-81ba-4155d9dbf76b

### CODE REVIEW

**Comment by dhruvmanila** [comment]
I think we might not want to include this in CI for now and only run it manually. This takes time because it requires loading VS Code, Ruff extension, initializing the language server.

File: src/test/diagnostics.test.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  });
+
+  assert.equal(actualDiagnostics.length, expectedDiagnostics.length);
+  actualDiagnostics.forEach((actualDiagnostic, i) => {
+    const expectedDiagnostic = expectedDiagnostics[i];
+    assert.equal(actualDiagnostic.message, expectedDiagnostic.message);
+    assert.deepEqual(actualDiagnostic.range, expectedDiagnostic.range);
+    assert.equal(actualDiagnostic.severity, expectedDiagnostic.severity);
+  });
+}
```
Comment (MichaReiser) [inline src/test/diagnostics.test.ts:43]: Could we use `assert.deepEqual(actualDiagnostics, expectedDiagnostics)`?

File: src/common/server.ts [hunk 2]
```diff
@@ -151,15 +151,14 @@ async function createNativeServer(
   serverName: string,
   outputChannel: LogOutputChannel,
   initializationOptions: IInitializationOptions,
-  ruffBinaryPath?: string,
-  ruffVersion?: VersionInfo,
+  ruffExecutable?: RuffExecutable,
```
Comment (MichaReiser) [inline src/common/server.ts:154]: Nice!

**Review by MichaReiser — APPROVED** [review]
I would find it worthwhile to spend 1-2h to figure out if we could port the CI workflow from the black extension. https://github.com/microsoft/vscode-black-formatter/blob/458ed8f03f03c6e021361b7dcf4a8c7ab7c3d5fb/.github/workflows/pr-check.yml#L91-L147 Not as part of this PR, but as a possible follow up

**Comment by dhruvmanila** [comment]
> I would find it worthwhile to spend 1-2h to figure out if we could port the CI workflow from the black extension. [microsoft/vscode-black-formatter@`458ed8f`/.github/workflows/pr-check.yml#L91-L147](https://github.com/microsoft/vscode-black-formatter/blob/458ed8f03f03c6e021361b7dcf4a8c7ab7c3d5fb/.github/workflows/pr-check.yml#L91-L147) Not as part of this PR, but as a possible follow up

Yeah, I'll look at it. I've one more reference (`vscode-python` extension) which I'll look at.

**Pull request merged**